File: port_forward.py
import os
import signal
import socket
import subprocess
import time


class PortForwardManager:
    """Manage a kubectl port-forward subprocess and expose a localhost endpoint.

    The port-forward process is started in its own process group so it can be
    cleanly terminated. This class is intentionally small and self-contained.
    """

    def __init__(self, kubectl: str, logger) -> None:
        self._kubectl = kubectl
        self._logger = logger
        self._proc: subprocess.Popen[str] | None = None
        self._local_port: int | None = None

    def _find_free_local_port(self, preferred_port: int = 8443) -> int:
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            s.bind(("127.0.0.1", preferred_port))
            s.close()
            return preferred_port
        except OSError:
            s.close()
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s2:
                s2.bind(("127.0.0.1", 0))
                return int(s2.getsockname()[1])

    def _wait_tcp_ready(self, host: str, port: int, timeout_seconds: float = 10.0) -> bool:
        start = time.time()
        while time.time() - start < timeout_seconds:
            try:
                with socket.create_connection((host, port), timeout=0.3):
                    return True
            except OSError:
                time.sleep(0.2)
        return False

    def start(
        self,
        namespace: str,
        service_name: str,
        remote_port: int = 443,
        preferred_local_port: int = 8443,
    ) -> str:
        """Start port-forward to the given service; returns "127.0.0.1:<port>".

        If already running, this is a no-op and returns the existing endpoint.
        """
        if self._proc and self._proc.poll() is None and self._local_port:
            return f"127.0.0.1:{self._local_port}"

        local_port = self._find_free_local_port(preferred_local_port)
        proc = subprocess.Popen(
            [
                self._kubectl,
                "-n",
                namespace,
                "port-forward",
                f"svc/{service_name}",
                f"{local_port}:{remote_port}",
                "--address",
                "127.0.0.1",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            text=True,
            start_new_session=True,
        )

        if not self._wait_tcp_ready("127.0.0.1", local_port, timeout_seconds=20):
            try:
                os.killpg(proc.pid, signal.SIGTERM)
            except Exception as e:
                self._logger.print_with_timestamp(f"Port-forward TERM failed: {e}", style="dim yellow")
            try:
                proc.wait(timeout=2)
            except Exception:
                try:
                    os.killpg(proc.pid, signal.SIGKILL)
                except Exception as e2:
                    self._logger.print_with_timestamp(f"Port-forward KILL failed: {e2}", style="dim yellow")
            raise RuntimeError("Port-forward to Central did not become ready")

        self._proc = proc
        self._local_port = local_port
        endpoint = f"127.0.0.1:{local_port}"
        self._logger.print_with_timestamp(f"✓ Port-forward active at https://{endpoint}", style="bold green")
        return endpoint

    def stop(self) -> None:
        if not self._proc:
            return
        try:
            if self._proc.poll() is None:
                try:
                    os.killpg(self._proc.pid, signal.SIGTERM)
                except Exception as e:
                    self._logger.print_with_timestamp(f"Port-forward TERM failed: {e}", style="dim yellow")
                try:
                    self._proc.wait(timeout=2)
                except Exception:
                    try:
                        os.killpg(self._proc.pid, signal.SIGKILL)
                    except Exception as e2:
                        self._logger.print_with_timestamp(f"Port-forward KILL failed: {e2}", style="dim yellow")
        finally:
            self._proc = None
            self._local_port = None

    def is_running(self) -> bool:
        return bool(self._proc and self._proc.poll() is None)

    def get_local_endpoint(self) -> str | None:
        if self._local_port and self.is_running():
            return f"127.0.0.1:{self._local_port}"
        return None

File: pkg/portforward/portforward.go
package portforward

import (
	"fmt"
	"net"
	"os/exec"
	"syscall"
	"time"

	"github.com/stackrox/roxie-golang/pkg/logger"
)

// Manager manages a kubectl port-forward subprocess and exposes a localhost endpoint
type Manager struct {
	kubectl   string
	logger    *logger.Logger
	proc      *exec.Cmd
	localPort int
}

// New creates a new PortForwardManager
func New(kubectl string, log *logger.Logger) *Manager {
	return &Manager{
		kubectl:   kubectl,
		logger:    log,
		localPort: 0,
	}
}

// findFreeLocalPort finds an available local port, preferring the given port
func (m *Manager) findFreeLocalPort(preferredPort int) (int, error) {
	// Try preferred port first
	ln, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", preferredPort))
	if err == nil {
		port := ln.Addr().(*net.TCPAddr).Port
		ln.Close()
		return port, nil
	}

	// Fall back to any available port
	ln, err = net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return 0, err
	}
	port := ln.Addr().(*net.TCPAddr).Port
	ln.Close()
	return port, nil
}

// waitTCPReady waits for a TCP port to become ready
func (m *Manager) waitTCPReady(host string, port int, timeoutSeconds float64) bool {
	timeout := time.Duration(timeoutSeconds * float64(time.Second))
	deadline := time.Now().Add(timeout)

	for time.Now().Before(deadline) {
		conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", host, port), 300*time.Millisecond)
		if err == nil {
			conn.Close()
			return true
		}
		time.Sleep(200 * time.Millisecond)
	}

	return false
}

// Start starts port-forward to the given service; returns "127.0.0.1:<port>"
// If already running, this is a no-op and returns the existing endpoint
func (m *Manager) Start(namespace, serviceName string, remotePort, preferredLocalPort int) (string, error) {
	// If already running, return existing endpoint
	if m.proc != nil && m.proc.Process != nil && m.localPort != 0 {
		return fmt.Sprintf("127.0.0.1:%d", m.localPort), nil
	}

	localPort, err := m.findFreeLocalPort(preferredLocalPort)
	if err != nil {
		return "", fmt.Errorf("failed to find free port: %w", err)
	}

	cmd := exec.Command(
		m.kubectl,
		"-n", namespace,
		"port-forward",
		fmt.Sprintf("svc/%s", serviceName),
		fmt.Sprintf("%d:%d", localPort, remotePort),
		"--address", "127.0.0.1",
	)

	// Set up process group for clean termination
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Setpgid: true,
	}

	cmd.Stdout = nil // Suppress output
	cmd.Stderr = nil

	if err := cmd.Start(); err != nil {
		return "", fmt.Errorf("failed to start port-forward: %w", err)
	}

	// Wait for port to be ready
	if !m.waitTCPReady("127.0.0.1", localPort, 20.0) {
		// Kill the process group
		if cmd.Process != nil {
			pgid, err := syscall.Getpgid(cmd.Process.Pid)
			if err == nil {
				syscall.Kill(-pgid, syscall.SIGTERM)
			}
			cmd.Wait()
		}
		return "", fmt.Errorf("port-forward did not become ready")
	}

	m.proc = cmd
	m.localPort = localPort
	endpoint := fmt.Sprintf("127.0.0.1:%d", localPort)
	m.logger.Successf("✓ Port-forward active at https://%s", endpoint)

	return endpoint, nil
}

// Stop stops the active port-forward if running
func (m *Manager) Stop() {
	if m.proc == nil || m.proc.Process == nil {
		return
	}

	// Try graceful termination first
	pgid, err := syscall.Getpgid(m.proc.Process.Pid)
	if err == nil {
		syscall.Kill(-pgid, syscall.SIGTERM)

		// Wait up to 2 seconds for graceful shutdown
		done := make(chan error, 1)
		go func() {
			done <- m.proc.Wait()
		}()

		select {
		case <-done:
			// Process terminated gracefully
		case <-time.After(2 * time.Second):
			// Force kill
			syscall.Kill(-pgid, syscall.SIGKILL)
		}
	}

	m.proc = nil
	m.localPort = 0
}
